code/controllers/subsystem/processing/overlays.dm

```diff
@@ -0,0 +1,160 @@
+
+/datum/subsystem/processing/overlays/Initialize()
+	initialized = TRUE
+	for(var/I in processing)
+		var/atom/A = I
+		A.compile_overlays()
+		CHECK_TICK
+	..()
+
+/datum/subsystem/processing/overlays/Recover()
+	overlay_icon_state_caches = SSoverlays.overlay_icon_state_caches
+	processing = SSoverlays.processing
+
+/datum/subsystem/processing/overlays/fire()
+	while(processing.len)
+		var/atom/thing = processing[processing.len]
+		processing.len--
+		if(thing)
+			thing.compile_overlays(FALSE)
+		if(MC_TICK_CHECK)
+			break
+	can_fire = processing.len > 0
+
+/atom/proc/compile_overlays()
+	if(LAZYLEN(priority_overlays) && LAZYLEN(our_overlays))
+		overlays = our_overlays + priority_overlays
+	else if(LAZYLEN(our_overlays))
+		overlays = our_overlays
+	else if(LAZYLEN(priority_overlays))
+		overlays = priority_overlays
+	else
+		overlays.Cut()
+	flags &= ~OVERLAY_QUEUED
+
+/atom/proc/extract_overlay_appearance(image_atom_or_string)
+	if (istext(image_atom_or_string))
+		var/static/image/stringbro = new()
+		var/list/icon_states_cache = SSoverlays.overlay_icon_state_caches 
+		var/list/cached_icon = icon_states_cache[icon]
+		if (cached_icon)
+			var/cached_appearance = cached_icon["[image_atom_or_string]"]
+			if (cached_appearance)
+				return cached_appearance
+		stringbro.icon = icon
+		stringbro.icon_state = image_atom_or_string
+		if (!cached_icon) //not using the macro to save an associated lookup
+			cached_icon = list()
+			icon_states_cache[icon] = cached_icon
+		var/cached_appearance = stringbro.appearance
+		cached_icon["[image_atom_or_string]"] = cached_appearance
+		return cached_appearance
+	else if (istype(image_atom_or_string, /image) || istype(image_atom_or_string, /atom))
+		var/image/I = image_atom_or_string
+		return I.appearance
+#if DM_VERSION >= 511
+	else if (istype(image_atom_or_string, /mutable_appearance/))
+		var/mutable_appearance/MA = image_atom_or_string
+		return MA.appearance
+#endif
+
+#define NOT_QUEUED_ALREADY (!(flags & OVERLAY_QUEUED))
+#define QUEUE_FOR_COMPILE flags |= OVERLAY_QUEUED; SSoverlays.can_fire = TRUE; if(SSoverlays.initialized) { SSoverlays.processing += src; } else { SSoverlays.processing[src] = src; } 
+/atom/proc/cut_overlays(priority = FALSE)
+	var/list/cached_overlays = our_overlays
+	var/list/cached_priority = priority_overlays
+	
+	var/need_compile = FALSE
+
+	if(LAZYLEN(cached_overlays)) //don't queue empty lists, don't cut priority overlays
+		cached_overlays.Cut()  //clear regular overlays
+		need_compile = TRUE
+
+	if(priority && LAZYLEN(cached_priority))
+		cached_priority.Cut()
+		need_compile = TRUE
+
+	if(NOT_QUEUED_ALREADY && need_compile)
+		QUEUE_FOR_COMPILE
+
+/atom/proc/cut_overlay(image, priority)
+	if(!image)
+		return
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = LAZYLEN(cached_overlays)
+	var/init_p_len = LAZYLEN(cached_priority)  //starter pokemon
+	
+	image = extract_overlay_appearance(image)
```

Comment:
this needs to support receiving a list.

Remember, you are taking a system that used to be done in code as list operations, something that supports a list on the right side, and moving it to encapsulated procs, it makes zero sense to not have the new system support lists, especially for proc overhead reason.